conf/schemas.go

```diff
@@ -58,50 +91,114 @@ func ReadSchemas(file string) (Schemas, error) {
 		}
 
 		if sec.ValueOf("pattern") == "" {
-			return nil, fmt.Errorf("[%s]: empty pattern", schema.Name)
+			return Schemas{}, fmt.Errorf("[%s]: empty pattern", schema.Name)
 		}
 		schema.Pattern, err = regexp.Compile(sec.ValueOf("pattern"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
 		}
 
 		schema.Retentions, err = ParseRetentions(sec.ValueOf("retentions"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
 		}
 
 		p := int64(0)
 		if sec.ValueOf("priority") != "" {
 			p, err = strconv.ParseInt(sec.ValueOf("priority"), 10, 0)
 			if err != nil {
-				return nil, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
+				return Schemas{}, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
 			}
 		}
 		schema.Priority = int64(p)<<32 - int64(i) // to sort records with same priority by position in file
 
 		schemas = append(schemas, schema)
 	}
 
-	sort.Sort(schemas)
-	return schemas, nil
+	sort.Sort(SchemaSlice(schemas))
+
+	return NewSchemas(schemas), nil
 }
 
 // Match returns the correct schema setting for the given metric
 // it can always find a valid setting, because there's a default catch all
 // also returns the index of the setting, to efficiently reference it
-func (s Schemas) Match(metric string) (uint16, Schema) {
-	for i, schema := range s {
+func (s Schemas) Match(metric string, interval int) (uint16, Schema) {
```

Comment:
I find this new function hard to understand.
One reason I think is because it's a mashup between two mental models:
A) all our schemas/retentions as 1 long list
B) a list of entries (based on pattern), and then each entry has a list of retention lists (due to the new "permutation" stuff). i.e. a nested structure.

This function has a few places where the logic works in the domain of B (e.g. taking the first retention of a given pattern, then working with `len(schema.Retentions)`, or iterating that list which actually represents items further down in the index list. I drafted some refactoring, but couldn't come up with anything better. I think a nested, B-style model would be nicer for our index datastructure, but then we would need two identifiers to represent a schema, which is to be avoided.